test/tests/functional/pbs_sched_rerun.py

```diff
@@ -0,0 +1,119 @@
+# the OpenPBS software.  For a copy of the commercial license terms and
+# conditions, go to: (http://www.pbspro.com/agreement.html) or contact the
+# Altair Legal Department.
+#
+# Altair's dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of OpenPBS and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair's trademarks, including but not limited to "PBS™",
+# "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+# subject to Altair's trademark licensing policies.
+
+
+from tests.functional import *
+
+
+hook_body = """
+import pbs
+e = pbs.event()
+if e.type == pbs.EXECJOB_BEGIN:
+    e.job.resources_used["foo"] = 123
+    if e.job.run_count == 1:
+        for v in e.vnode_list:
+            if (v == "%s"):
+                e.vnode_list[v].state = pbs.ND_OFFLINE
+        e.job.rerun()
+        e.reject("reruen job")
+e.accept()
+"""
+
+
+class TestSchedRerun(TestFunctional):
+    """
+    Tests to verify scheduling of rerun jobs.
+    """
+
+    @requirements(num_moms=2)
+    def test_rerun_job_over_reservation(self):
+        """
+        Test that job will not run over reservation/top job
+        after first failed attempt to run (with used resource set).
+        """
+        now = int(time.time())
+
+        usage_string = 'test requires two moms,' + \
+                       'use -p "servers=M1,moms=M1:M2"'
+
+        if len(self.moms.values()) != 2:
+            self.skip_test(usage_string)
+
+        self.momA = self.moms.values()[0]
+        self.momB = self.moms.values()[1]
+        self.hostA = self.momA.shortname
+        self.hostB = self.momB.shortname
+
+        if not self.hostA or not self.hostB:
+            self.skip_test(usage_string)
+
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'managers': (INCR, '%s@*' % TEST_USER)})
+
+        a = {'type': 'long', 'flag': 'i'}
+        r = 'foo'
+        self.server.manager(MGR_CMD_CREATE, RSC, a, id=r)
+
+        hook_name = "rerun_job"
+        a = {'event': 'execjob_begin',
+             'enabled': 'True'}
+        rv = self.server.create_import_hook(
+            hook_name, a, hook_body % self.hostA, overwrite=True)
+        self.assertTrue(rv)
+
+        a = {'reserve_start': now + 60,
+             'reserve_end': now + 7200}
+        h = [self.momB.shortname]
+        r = Reservation(TEST_USER, attrs=a, hosts=h)
+
+        self.server.submit(r)
+
+        a = {'Resource_List.select': '1:ncpus=1',
+             'Resource_List.walltime': '24:00:00'}
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(1000)
+        jid = self.server.submit(j)
+
+        self.logger.info("Wait for the job to try to rerun (10 seconds)")
+        time.sleep(10)
+
+        # the job shoud not run because first mom is offline
+        # second mom is occupied by maintenance reservation
```

Comment:
The reservation is not a maintenance reservation.  Also, since you don't request a node for the reservation, it can be on either node.

Reply:
It can not. It is the maintenance reservation. In the following lines, the `hosts` is requested. See:
```
        h = [self.momB.shortname]
        r = Reservation(TEST_USER, attrs=a, hosts=h)
```
I will change the first line of the code snippet to `h = [self.hostB]` to make it nicer.

See the ptl log. Here we can make sure the maintenance reservation is confirmed and a host is requested. The host is `self.hostB`:
```
2024-03-14 13:34:34,621 INFOCLI  rocky9(run_cmd): sudo -H -u pbsuser /opt/pbs/bin/pbs_rsub -R 202403141335.32 -E 202403141534.32 --hosts almalinux
2024-03-14 13:34:34,846 INFO     submit to rocky9 as pbsuser: reservation M9.rocky9 CONFIRMED OrderedDict([('reserve_start', 1710419732), ('reserve_end', 1710426872)])
```

Reply:
Ahh, missed the hosts part.  Indeed it's a maintenance reservation.